### DISCUSSION TITLE
Gas Fees

### DISCUSSION THREAD

**Discussion by Jpja** [CounterpartyXCP/CIPs#123] [General]
One Counterparty action requires one Bitcoin transaction. This is the general logic, and the miner fee is the de-facto "gas".

There are several special cases where the BTC miner fee is not sufficient. In these cases some additional XCP fee is deducted automatically. If the XCP balance is too low, the action is not executed and deemed "invalid".

**Dividend**
A token is distributed to all holder of another token. This is done through one CNTRPRTY message, i.e. only one BTC tx fee. To prevent spam, two rules were added. 1) Only the _asset owner_ can pay a distribution. 2) An XCP fee is charged for each recipient.

**Multi send**
Multi-Peer Multi-Asset sends ([MPMA](https://github.com/CounterpartyXCP/cips/blob/master/cip-0010.md)) enables batch sending of multiple tokens to multiple recipients. This is done through one CNTRPRTY message where LUT tables are used to compress the asset IDs and recipient addresses. The message is typically >80 bytes and [P2SH](https://github.com/CounterpartyXCP/cips/blob/master/cip-0006.md) encoding is used. This requires a chain of two bitcoin transactions and a relatively expensive BTC fee. Still, it is significantly cheaper than making N normal sends through N btc transactions, so additional XCP fees are charged.

**Sweep**
Like dividend, one short CNTRPRTY message lets you transfer all token balances and/or asset ownerships. Thus XCP fee.

**Named asset registration**
To protect the namespace, a 0.5 XCP fee is charged for registering a unique name.

**Sub-asset registration**
A 0.25 XCP fee is charged. The owner of the parent asset has the sole right of issuing sub-assets, so there is no direct "protect the namespace" logic here. Except that a sub-asset uses the _numeric_ namespace, which also should have some protection.

**Numeric asset registration**
No XCP fee (inconsistent with the sub-asset fee). [CIP 29](https://github.com/CounterpartyXCP/cips/blob/master/cip-0029.md) proposes a 0.10 XCP fee for registering a new numeric name, and 0.01 for all updates to any asset (named, sub or numeric).

**Dispensers**
No XCP fee. I think a fee should be introduced for several reasons. 1) No CNTRPRTY message to trigger a dispense. Dispensers like ORDIPEPE, OXBT and OGPASS are triggered by bitcoin transactions with no Counterparty action intended. Even if this won't be possible with the `no_history` rule updated, dispensers can still be set up to be activated by regular bitcoin transactions. Like deposit to exchange, get s**tcoin in return. 2) Bundle dispensers. One address can set up a 2,3,..N dispensers. One incoming BTC tx will trigger all N dispensers. 3) If [batch-sending](https://github.com/CounterpartyXCP/counterparty-lib/issues/1148) is introduced (currently only the first output can trigger a dispense, this will change so that all outputs can trigger a dispense). This lowers the marginal cost from  141 bytes of BTC fees to 31 bytes,  or by 78%! Batch-sending to bundle dispensers will quickly bloat the Counterparty DB.

To generalize, the **reasons for XCP fees are**:
1) bulk actions where the BTC tx fee is insufficient
2) protect the namespace 
3) relative pricing. This is not used so far, but was one of the arguments for adding a fee to numeric assets. A project, SRC20, issued zero-issuance numeric assets. Broadcasts would be easier on the Counterparty DB, therefore an XCP fee on issuance would incentivize using broadcasts instead. 

A good development is taking place with the upgrade to [Sweep fees](https://github.com/CounterpartyXCP/counterparty-lib/issues/1220#issuecomment-1747037727). Different actions are given `DB hit` scores. Updating a balance is 2 db hits. An issuance is 4 db hits.

The cost of a db hit is set at $0.003 = XCP 0.001. The "social contract" could be that at every update the XCP price parameter is changed to the current price. Everything else stays. 

To eliminate the need of adjusting the XCP price manually, a dynamic adjustment should be considered. Base it on the DEX XCP/BTC price (not USD). This must be researched thoroughly (I see risks of price manipulation).

For dispensers I would like a similar fee calculation added. The XCP fee is escrowed from the dispenser operator at setup, and based on the number of potential dispenses. Deducted once the dispenser is emptied, or refunded accordingly if the dispenser gets closed.

_Your thoughts?_

**Comment by jotapea** [comment]
I lean more into removing the need for XCP for as much as possible. So, I would not add it as a requirement for dispensers. And I would also remove it for sub-assets, to be consistent with numeric ones while showing the benefits of having a "root" asset name.

**Reply by Jpja** [reply]
Do you agree on introducing batch-sending to dispensers? IMO we can have XCP gas fees or bundle/batch dispensers but not both.

**Reply by jotapea** [reply]
Without consideration of how checking all payment outputs will affect sync performance, I think it is an improvement to detect dispensers for all outputs. So I agree.

I'm not sure how adding XCP fees improves anything here.
